Sources/SwiftCompilerPlugin/PluginMacroExpansionContext.swift

```diff
@@ -0,0 +1,253 @@
+    return SourcePosition(
+      fileName: base.location.fileName,
+      utf8Offset: localPosition.advanced(by: positionOffset).utf8Offset
+    )
+  }
+
+  /// Get position range of `node` in the known root nodes.
+  /// The root node of `node` must be one of the retured value from `add(_:)`.
+  func range(
+    of node: Syntax,
+    from startKind: PositionInSyntaxNode = .afterLeadingTrivia,
+    to endKind: PositionInSyntaxNode = .beforeTrailingTrivia
+  ) -> SourceRange? {
+    guard let base = self.knownSourceSyntax[node.root.id] else {
+      return nil
+    }
+    let localStartPosition = node.position(at: startKind)
+    let localEndPosition = node.position(at: endKind)
+    assert(localStartPosition <= localEndPosition)
+
+    let positionOffset = base.location.offset
+
+    return SourceRange(
+      fileName: base.location.fileName,
+      startUTF8Offset: localStartPosition.advanced(by: positionOffset).utf8Offset,
+      endUTF8Offset: localEndPosition.advanced(by: positionOffset).utf8Offset
+    )
+  }
+
+  /// Get location of `node` in the known root nodes.
+  /// The root node of `node` must be one of the retured value from `add(_:)`.
+  func location(of node: Syntax, at kind: PositionInSyntaxNode, filePathMode: SourceLocationFilePathMode) -> SourceLocation? {
+    guard let base = self.knownSourceSyntax[node.root.id] else {
+      return nil
+    }
+    let file: String
+    switch filePathMode {
+    case .fileID: file = base.location.fileID
+    case .filePath: file = base.location.fileName
+    }
+
+    let localPosition = node.position(at: kind)
+    let localLocation = base.locationConverter.location(for: localPosition)
+
+    let positionOffset = base.location.offset
+    let lineOffset = base.location.line - 1
+    let columnOffset = localLocation.line == 1 ? base.location.column : 0
+
+    return SourceLocation(
+      line: localLocation.line! + lineOffset,
+      column: localLocation.column! + columnOffset,
+      offset: localLocation.offset + positionOffset,
+      file: file
+    )
+  }
+}
+
+fileprivate extension Syntax {
+  /// get a position in the node by `PositionInSyntaxNode`.
+  func position(at pos: PositionInSyntaxNode) -> AbsolutePosition {
+    switch pos {
+    case .beforeLeadingTrivia:
+      return self.position
+    case .afterLeadingTrivia:
+      return self.positionAfterSkippingLeadingTrivia
+    case .beforeTrailingTrivia:
+      return self.endPositionBeforeTrailingTrivia
+    case .afterTrailingTrivia:
+      return self.endPosition
+    }
+  }
+}
+
+class PluginMacroExpansionContext {
+  /// The macro expansion discriminator, which is used to form unique names
+  /// when requested.
+  ///
+  /// The expansion discriminator is combined with the `uniqueNames` counters
+  /// to produce unique names.
+  private var discriminator: String
+
+  private var sourceManger: SourceManager
+
+  /// Counter for each of the uniqued names.
+  ///
+  /// Used in conjunction with `expansionDiscriminator`.
```

Comment:
I think `expansionDiscriminator` no longer exists.

Reply:
Renamed `discriminator` to `expansionDiscriminator` to align with ASTGen.